ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java

```diff
@@ -372,12 +385,87 @@ protected void updateResolvedContainers(Account account, Collection<Container> r
     }
 
     // write added/modified containers to in-memory cache
-    infoMapLock.writeLock().lock();
-    try {
-      accountInfoMapRef.get().addOrUpdateContainers(resolvedContainers);
-    } finally {
-      infoMapLock.writeLock().unlock();
+    updateContainersInCache(resolvedContainers);
+  }
+
+  /**
+   * Gets the {@link Container} by its name and parent {@link Account} name by looking up in in-memory cache.
+   * If it is not present in in-memory cache, it queries from mysql db and updates the cache.
+   * @param accountName the name of account which container belongs to.
+   * @param containerName the name of container to get.
+   * @return {@link Container} if found in cache or mysql db. Else, returns {@code null}.
+   * @throws AccountServiceException
+   */
+  @Override
+  public Container getContainerByName(String accountName, String containerName) throws AccountServiceException {
+    Container container = null;
+
+    if (recentNotFoundContainersCache.contains(accountName + SEPARATOR + containerName)) {
```

Comment:
Imagine a case where customers attempt to write into a non-existent container and get an error. They immediately create a new container (with same account/container name). If the request hits the same frontend, do we still return null?

Reply:
That's a good point.  I think every time we add a new container to the cache (create request of DB sync), we need to remove it (if needed) from the not-found cache.